url696d9a93c9_liaosiwei_bayesnetwork_tgz-pJ8-tutorial_TutorialJ8/src/com/alibaba/DataPlot.java

```diff
@@ -0,0 +1,50 @@
+package com.alibaba;
+
+import org.math.plot.*;
+
+import javax.swing.*;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * User: sweeliao@gmail.com
+ * Date: 14-3-21
+ * Time: 上午10:50
+ */
+public class DataPlot {
+    public static void main(String[] args) {
+        DataRead reader = new DataRead();
+        reader.read();
+        if (!reader.read()) {
+            System.out.println("read file error");
+            return;
+        }
+        Map<String, Map<String, Map<Integer, HashSet<Integer>>>> table = reader.getData();
+        reader.close();
+
+        double[] x = new double[table.size()];
+        double[] y = new double[table.size()];
+        double[] z = new double[table.size()];
+        Plot2DPanel plot = new Plot2DPanel();
+
+        int num = 0;
+        for (String user: table.keySet()) {
+            long click = 0, buy = 0, save = 0;
+            for (String brand: table.get(user).keySet()) {
+                Set<Integer> ii = table.get(user).get(brand).get(1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.